Rails 6 Quickstarts

davidpatrick opened the pull request:
- Migrates from secrets to config_for
- Adds non-interactive form post redirect
- Consolidates tutorials into one easy to use tutorial
- Migrates to Rails 6
- Points to new sample
- Move troubleshooting steps into a seperate tutorial

davidpatrick committed:
Rails 6 Quickstarts

- migrates from secrets to config_for
- adds non-interactive formpost redirect
- point to new sample
- move troubleshooting into a seperate tutorial

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -9,169 +9,124 @@ topics:
-:::
-
-Create a file named `auth0.rb` under `config/initializers` and configure the **OmniAuth** middleware in it.
+Create an initializer file in the root of your project's dierctory `./config/initializers/auth0.rb` and [configure](https://github.com/auth0/omniauth-auth0#additional-authentication-parameters) the **OmniAuth** middleware.
```
```suggestion
Create an initializer file in the root of your project's directory `./config/initializers/auth0.rb` and [configure](https://github.com/auth0/omniauth-auth0#additional-authentication-parameters) the **OmniAuth** middleware.
```

It is a bit confusing that it mentions to create a file in the root of your project, yet it mentions a path containing two sub folders.

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -9,169 +9,124 @@ topics:
-    Rails.application.secrets.auth0_domain,
+    Rails.application.config.auth0['auth0_client_id'],
+    Rails.application.config.auth0['auth0_client_secret'],
+    Rails.application.config.auth0['auth0_domain'],
```
Pardon my ignorance on ruby, but is there any reason why `auth0_config = Rails.application.config_for(:auth0)` is not being used (line 50)?

stevehobbsdev replied:
I had this comment too, feels like we can save some typing and reduce a bit of code.

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -9,169 +9,124 @@ topics:
-Rails uses a cookie by default to store session data. If you run into `CookieOverflow` error when trying to save the user info into the session, please read [the troubleshooting section](#actiondispatch-cookies-cookieoverflow) on how to store session data in memory instead.
-:::
+### Add an Auth0 controller
+Create an Auth0 controller file in the root of your project's dierctory `./config/controllers/auth0_controller.rb` for handling of the authentication callback.  Inside of the callback method assign the hash of user information returned `request.env['omniauth.auth']` to the active session.
```
```suggestion
Create an Auth0 controller file in the root of your project's directory `./config/controllers/auth0_controller.rb` for handling of the authentication callback.  Inside of the callback method assign the hash of user information, returned as `request.env['omniauth.auth']`, to the active session.
```

Same here, it mentions root directory yet places a file two sub-levels deep.

stevehobbsdev replied:
Feels like an opportunity here to include a Rails `generate` command that the reader can copy & paste to create this controller.

```
rails generate controller auth0 callback failure --no-assets --no-helper
```

WDTY? We do this later for Logout, for example.

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
````diff
@@ -9,169 +9,124 @@ topics:
-```
-<!-- app/views/home/show.html.erb -->
+### Redirect User to Login
+In order for us to redirect the user to login without any user interaction, we will need to create a non-interactive redirect page.  The redirect page will [prevent forged authentication requests](https://github.com/cookpad/omniauth-rails_csrf_protection) and handle the `:POST` request for us automatically using javascript.
````
```suggestion
In order for us to redirect the user to login without any user interaction, we will need to create a non-interactive redirect page. The redirect page will [prevent forged authentication requests](https://github.com/cookpad/omniauth-rails_csrf_protection) and handle the `:POST` request for us automatically using javascript.
```

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -9,169 +9,124 @@ topics:
-
-You can use a controller `concern` to control access to routes that require the user to be authenticated:
+## Display User Profile
+To display the user's profile, your application should provide a protected route.  You can use a controller `concern` to control access to routes.  Create the following file `./app/controllers/concerns/secured.rb`
```
```suggestion
To display the user's profile, your application should provide a protected route. You can use a controller `concern` to control access to routes. Create the following file `./app/controllers/concerns/secured.rb`
```

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -180,72 +135,35 @@ module Secured
-
-Include the `concern` in this new controller to prevent unauthenticated users from accessing its routes:
-
+After you have created the secured controller concern, include it in any controllers requires a logged in user.
```
```suggestion
After you have created the secured controller concern, include it in any controller that requires a logged in user.
```

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -180,72 +135,35 @@ module Secured
-  new_path = "#{env['SCRIPT_NAME']}#{OmniAuth.config.path_prefix}/failure?message=#{message_key}&error_description=#{error_description}"
-  Rack::Response.new(['302 Moved'], 302, 'Location' => new_path).finish
-}
+Here is an example of displaying that information on the page
```
```suggestion
Here is an example of displaying the corresponding user's information
```

frederikprijck commented on articles/quickstart/webapp/rails/01-login.md:
```diff
@@ -273,8 +191,8 @@ In `logout_helper.rb` file add the methods to generate the logout URL.
-    domain = Rails.application.secrets.auth0_domain
-    client_id = Rails.application.secrets.auth0_client_id
+    domain = Rails.application.config.auth0.auth0_domain
+    client_id = Rails.application.config.auth0.auth0_client_id
```
Above, the following syntax was used `Rails.application.config.auth0['auth0_client_id'],`, should we stick to one syntax or is there any reason these use a different way to access the properties?

stevehobbsdev replied:
I think we should use `auth0_config = Rails.application.config_for(:auth0)` here again and index into that instead of `Rails.application.config.auth0`

frederikprijck commented on articles/quickstart/webapp/rails/02-troubleshooting.md:
```diff
@@ -0,0 +1,105 @@
+Restart your Rails server for these changes to take effect.
+
+:::note
+It is recommended that a memory store such as MemCached being used for production applications.
```
```suggestion
It is recommended that a memory store, such as MemCached, is being used for production applications.
```

stevehobbsdev commented on articles/quickstart/webapp/rails/01-login.md:
````diff
@@ -9,169 +9,124 @@ topics:
-```
-<!-- app/views/home/show.html.erb -->
+### Redirect User to Login
+In order for us to redirect the user to login without any user interaction, we will need to create a non-interactive redirect page.  The redirect page will [prevent forged authentication requests](https://github.com/cookpad/omniauth-rails_csrf_protection) and handle the `:POST` request for us automatically using javascript.
````
What's missing from here for me is the reasoning behind _why_ a user might want to do this. We could include a sentence, something like "As an example, you might want to do this when..." just to give an example scenario when this might be useful.

Right now without it, it's a little abstract.

stevehobbsdev replied:
Also in general it's a good approach, but the only annoyance I found was that if a user decides to back out of the login page back to their app, they'll hit the redirect page which then automatically forwards them to the login page again. They have to do that thing where they click back twice to get to the right page.

davidpatrick replied:
@stevehobbsdev if they were to back out of the login page to a protected route, wouldnt they be redirected back to login traditionally?

davidpatrick replied:
I think we will move this out of the quickstart and into an example

stevehobbsdev reviewed (changes requested):
I've run through it and (other than what I've commented below) I found a couple of issues. I understand that they might not be part of this diff but I wonder if they're worth fixing up all the same.

Bear in mind for my run through I am using Rails 6.1.2.1 and Ruby 2.7.2, both of which look to be permitted by the system requirements on the doc.

### Can't get the session to "stick"

Once I add the `Secured` concern, I get stuck in a redirect loop, as after login I get redirected to `/dashboard`, which checks that I am logged in and, if not, I am redirected to login again. But `session[:userinfo]` comes out as `nil`. I've checked in `auth0#callback` and it's getting a value and setting the session, but I might be missing something else. Any idea?

### Logout URL

Really minor one but if you follow the quickstart you get a logout URL of `/logout/logout` - would it be better to have just `/logout`?

❗ When generating the logout URL, `URI::HTTPS.build` didn't work for me, says `build` is undefined. This could be a Ruby version thing, we should make sure the system requirements are correct.

### Requirements

On the requirements, the current live one says "Ruby 2.3.1+ | Rails 5.0+ or Rails 4.2+" and the new one says "Ruby 4.2+". First off, I guess the new one should be "Rails 4.2+" (not Ruby), but also are we sure it works between 4.2 and 5? Do we know why the old one called out 4.2 and 5.0 separately, or was that just an oversight?

### Headings

On other more recent quickstarts (e.g. [React](https://auth0.com/docs/quickstart/spa/react)) we used headings like "Add login to your Application" and "Add Logout to your Application", should we do the same here?

davidpatrick committed:
Apply suggestions from code review

Co-authored-by: Frederik Prijck <frederik.prijck@auth0.com>

davidpatrick committed:
Applying feedback

davidpatrick committed:
Updates

davidpatrick committed:
Applying feedback

davidpatrick commented:
@stevehobbsdev 

> Really minor one but if you follow the quickstart you get a logout URL of /logout/logout - would it be better to have just /logout?

Fixed instructions for this.  Removed the LogoutController, and added Logout to Auth0Controller

> When generating the logout URL, URI::HTTPS.build didn't work for me, says build is undefined. This could be a Ruby version thing, we should make sure the system requirements are correct
Not sure whats going on here, but `URI::HTTPS.build` works and is available for me on ruby 2.7.2 and/or 2.5.x/2.6.x/3.x.  

One thing I did run into when starting from scratch was that `root_url` was empty.  I've put a code comment on this variable to warn user that `root to:` needs to be set in the routes file.

> On the requirements, the current live one says "Ruby 2.3.1+ | Rails 5.0+ or Rails 4.2+" and the new one says "Ruby 4.2+". First off, I guess the new one should be "Rails 4.2+" (not Ruby), but also are we sure it works between 4.2 and 5? Do we know why the old one called out 4.2 and 5.0 separately, or was that just an oversight?

Cleaned this up

> On other more recent quickstarts (e.g. React) we used headings like "Add login to your Application" and "Add Logout to your Application", should we do the same here?

Updated these to make consistent

> Once I add the Secured concern, I get stuck in a redirect loop, as after login I get redirected to /dashboard, which checks that I am logged in and, if not, I am redirected to login again. But session[:userinfo] comes out as nil. I've checked in auth0#callback and it's getting a value and setting the session, but I might be missing something else. Any idea?

I've removed the non-interactive post redirect from the secured concern for the quickstart.
